File: gastromanager-frontend/src/app/features/submenus/application/mappers/submenu.mapper.ts
import { SubmenuRequestDto } from "../../domain/models/submenu-request-dto.interface";
import {SubmenuSummaryResponseDto} from "../../domain/models/submenu-summary-response-dto.interface";
import {Submenu} from "../../domain/models/submenu.interface";

export function mapToSubmenu(dto: SubmenuSummaryResponseDto): Submenu {
  return {
    uuid: dto.uuid,
    name: dto.name,
    description: dto.description,
    menu: {
      uuid: dto.menuName, // Assuming menuName is the UUID of the menu
      name: dto.menuName, // Assuming menuName is also used as the name
      description: '', // Placeholder, as this information is not provided in the DTO
      isEnabled: true, // Placeholder, as this information is not provided in the DTO
    },
    createdAt: new Date(), // Placeholder, as this information is not provided in the DTO
    updatedAt: new Date(), // Placeholder, as this information is not provided in the DTO
  };
}

export function mapToSubmenuRequestDto(submenu: Submenu): SubmenuRequestDto {
  return {
    name: submenu.name,
    description: submenu.description,
    menuUuid: submenu.menu.uuid,
  }
}

File: gastromanager-backend/src/main/java/com/kaiho/gastromanager/application/submenu/mapper/SubmenuMapper.java
package com.kaiho.gastromanager.application.submenu.mapper;

import com.kaiho.gastromanager.application.submenu.dto.request.SubmenuRequestDto;
import com.kaiho.gastromanager.application.submenu.dto.request.SubmenuUpdateRequestDto;
import com.kaiho.gastromanager.application.submenu.dto.response.SubmenuDetailResponseDto;
import com.kaiho.gastromanager.application.submenu.dto.response.SubmenuSummaryResponseDto;
import com.kaiho.gastromanager.domain.menu.model.Menu;
import com.kaiho.gastromanager.domain.submenu.domain.Submenu;
import org.springframework.stereotype.Component;

@Component
public class SubmenuMapper {

    public Submenu toDomain(SubmenuRequestDto submenuRequestDto) {
        if (submenuRequestDto == null) {
            return null;
        }
        return Submenu.builder()
                      .name(submenuRequestDto.name())
                      .description(submenuRequestDto.description())
                      .menu(Menu.builder().uuid(submenuRequestDto.menuUuid()).build())
                      .isEnabled(true)
                      .build();
    }

    public SubmenuSummaryResponseDto toResponseSummary(Submenu submenu) {
        if (submenu == null) {
            return null;
        }
        return SubmenuSummaryResponseDto.builder()
                                        .uuid(submenu.getUuid())
                                        .name(submenu.getName())
                                        .description(submenu.getDescription())
                                        .menuName(submenu.getMenu().getName())
                                        .isEnabled(submenu.isEnabled())
                                        .build();
    }

    public SubmenuDetailResponseDto toResponseDetail(Submenu submenuById) {
        if (submenuById == null) {
            return null;
        }
        return SubmenuDetailResponseDto.builder()
                                       .uuid(submenuById.getUuid())
                                       .name(submenuById.getName())
                                       .description(submenuById.getDescription())
                                       .menuUuid(submenuById.getMenu().getUuid())
                                       .menuName(submenuById.getMenu().getName())
                                       .isEnabled(submenuById.isEnabled())
                                       .createdBy(submenuById.getCreatedBy())
                                       .createdDate(submenuById.getCreatedDate())
                                       .updatedBy(submenuById.getUpdatedBy())
                                       .updatedDate(submenuById.getUpdatedDate())
                                       .build();
    }

    public Submenu toDomain(SubmenuUpdateRequestDto submenuRequestDto) {
        if (submenuRequestDto == null) {
            return null;
        }
        return Submenu.builder()
                      .name(submenuRequestDto.name())
                      .description(submenuRequestDto.description())
                      .build();
    }
}
